### PR TITLE
mpd: implement albumart/readpicture commands

### PR DESCRIPTION
**Pull request by grobian**

### CODE REVIEW

**Review by ejurgensen — COMMENTED** [review]
Looks great. In fact impressive, contrary to many first PRs this gives me the impression you studied the existing code and made sure the changes would fit. 

I have two minor remarks and then a question: I thought clients woould only use readpicture if the protocol version number that OwnTone declares was increased (not sure how much), but that doesn't seem to be the case with your clients. Increasing the version number brings with it other issues, so I think that's why chme couldn't easily add readpicture. Do you know if I am right about this?

Btw since I'm away the next couple of weeks, I won't be able to merge during that time.

**Comment by grobian** [comment]
> Looks great. In fact impressive, contrary to many first PRs this gives me the impression you studied the existing code and made sure the changes would fit.

Thanks, but please be aware that the artwork server (port 6601) in the same file was of great inspiration – in fact I just copied/pasted bits from there.

> I have two minor remarks and then a question: I thought clients woould only use readpicture if the protocol version number that OwnTone declares was increased (not sure how much), but that doesn't seem to be the case with your clients. Increasing the version number brings with it other issues, so I think that's why chme couldn't easily add readpicture. Do you know if I am right about this?

I'm not very deep into the MPD protocol at this point.
Some background: I have a personal need to make some things "work".  That is, I have an MPD client, and I want to stream my music from anywhere.  So I need the httpd stream.  I can do this with MPD server, but it doesn't provide a nice frontend, or ability to stream to AirPlay/Chromecast devices, so instead of having both owntone and MPD, I'm trying to fill in the gaps I see in owntone's MPD support.  Gaps here are these two PRs (httpd plugin exposure, missing albumart) and at the moment the library shows up empty to the client I use which I still need to investigate.

Coming back to the protocol version.  I can try and see if I can find something about which version introduces what functionality.  However, it seems that there are enough clients (given my own experience and that of people in the feature request linked) that don't bother about that and just call the functions they hope are there, and silently ignore failures or something.  That said if there is a way in which I can contribute to owntone, I'm motivated at this point of working on its MPD support to work with the client I currently use.  Of course, then I can check a few other clients too on mac/linux.

> Btw since I'm away the next couple of weeks, I won't be able to merge during that time.

Running on Gentoo here, so keeping a patch around for a while is trivial :)  Thanks!

File: src/mpd.c [hunk 1]
```diff
@@ -2606,6 +2606,72 @@ mpd_command_save(struct evbuffer *evbuf, int argc, char **argv, char **errmsg, s
...
+
+  format = artwork_get_item(evbuffer, itemid,
+  			    ART_DEFAULT_WIDTH, ART_DEFAULT_HEIGHT, 0);
+  if (format < 0)
+    {
+      *errmsg = safe_asprintf("Document was not found");
+      evbuffer_free(evbuffer);
+      return ACK_ERROR_ARG;
+    }
+  else
```
[the remark this answers is not in the document]
Comment (grobian) [reply src/mpd.c:2648]: I admit, I copied/pasted here, and thought this was "style".

File: src/mpd.c [hunk 2]
```diff
@@ -2606,6 +2606,72 @@ mpd_command_save(struct evbuffer *evbuf, int argc, char **argv, char **errmsg, s
...
+
+	  default:
+	    type = "image/jpeg";
+	    break;
+	}
+    }
+
+  size = evbuffer_get_length(evbuffer);
+  evbuffer_add_printf(evbuf,
+      "size: %zd\n"
```
[the remark this answers is not in the document]
Comment (grobian) [reply src/mpd.c:2664]: ack

**Comment by grobian** [comment]
> @grobian Does it apply to radio streams? Some of them supply covers. Radio paradise is one example.

How would I test that?  Is that a stream?  Can you PM me the url/instructions so I can check what you're looking for?

**Comment by grobian** [comment]
Just tried this, doesn't work.  I see the art in the webui, it isn't relayed to the client.  Interesting to see how the webui calls it, perhaps there's a little tweak we can do to make it work.

**Comment by ejurgensen** [comment]
I think it is because db_file_xxx will only retrieve tracks from the library, but items added via the web ui are only in the queue. What is the exact content of the readpicture command?

**Comment by grobian** [comment]
The response is the binary png/jpeg bytes representing the picture.

**Comment by ejurgensen** [comment]
Yes of course. My question was about the request.

**Comment by grobian** [comment]
And indeed, there are two problems, because the albumart/readpicture commands have a file argument, which is likely whatever the current playing thing is, so we'd have to know it's not a local file, but stream, in order to know that the image needs to be retrieved from another place.

**Comment by grobian** [comment]
beware that most likely the best we can implement is that it show the picture of whatever was playing at the time the client first asked for the picture
once the stream changes a track, the client still sees the same stream, so no "new" song or anything, so it won't re-request albumart
that said it might make more sense to get some kind of radio logo up there for streams -- perhaps we can try and retrieve the favicon or something

**Comment by ejurgensen** [comment]
The mpd protocol supports sending events to clients. Maybe you can check if any of the MPD_IDLE_xxx events make clients call readpicture.

**Comment by grobian** [comment]
idle player perhaps would cause the art to be re-retrieved, as [the description](https://mpd.readthedocs.io/en/latest/protocol.html#id14) says:
player: the player has been started, stopped or seeked or tags of the currently playing song have changed (e.g. received from stream)

**Comment by grobian** [comment]
Just to make sure, this PR isn't good as it is, because it needs to return the binary data in chunks.  It currently sends one giant chunk, but it should respect a limit on this (which defaults to 8K).  The PR #1791 adds a configurable limit (by the client) and should be used here to send out the art in chunks as configured.

**Comment by grobian** [comment]
this is better done using https://github.com/grobian/owntone-server/commit/c2b439ad46e499509e81d1616af3904f87f1a00d and https://github.com/grobian/owntone-server/commit/ab79a20e08cef64e60d732092bb754d8458f951e

**Pull request closed without merging**